README
GPUJPEG
  JPEG encoder and decoder library and console application for NVIDIA GPUs.

AUTHOR:
  Martin Srom, CESNET z.s.p.o
  Jan Brothánek
  Martin Jirman
  Jiri Matela
  Martin Pulec
  Petr Holub

DESCRIPTION:
  The first test implementation of the JPEG image compression standard for 
NVIDIA GPUs used for real-time transmission of high-definition video.

OVERVIEW:
-It uses NVIDIA CUDA platform.
-Not optimized yet (it is only the first test implementation).
-Encoder and decoder use Huffman coder for entropy encoding/decoding.
-Encoder produces by default baseline JPEG codestream which consists of proper codestream
 headers and one scan for each color component without subsampling and it uses
 restart flags that allows fast parallel encoding. The quality of encoded 
 images can be specified by value 0-100.
-Optionally encoder can produce interleaved stream (all components in one scan) or/and
 subsampled stream.
-Decoder can decompress only JPEG codestreams that can be generated by encoder. If scan 
 contains restart flags, decoder can use parallelism for fast decoding.
-Encoding/Decoding of JPEG codestream is divided into following phases:
   Encoding:                       Decoding
   1) Input data loading           1) Input data loading
   2) Preprocessing                2) Parsing codestream 
   3) Forward DCT                  3) Huffman decoder
   4) Huffman encoder              4) Inverse DCT
   5) Formatting codestream        5) Postprocessing
 and they are implemented on CPU or/and GPU as follows:
   -CPU: 
      -Input data loading
      -Parsing codestream
      -Huffman encoder/decoder (when restart flags are disabled)
      -Output data formatting
   -GPU: 
      -Preprocessing/Postprocessing (color component parsing, 
       color transformation RGB <-> YCbCr)
      -Forward/Inverse DCT (discrete cosine transform)
      -Huffman encoder/decoder (when restart flags are enabled)  

PERFORMANCE:
  Following tables summarizes encoding/decoding performance using NVIDIA 
GTX 580 for non-interleaved and non-subsampled stream with different quality 
settings (time, PSNR and encoded size values are averages of encoding several 
images, each of them multiple times):

Encoding:
         |           4k (4096x2160)         |         HD (1920x1080)
 --------+----------------------------------+---------------------------------
 quality | duration |     psnr |       size | duration |     psnr |       size
 --------+----------+----------+------------+---------------------------------
      10 | 26.79 ms | 29.33 dB |  539.30 kB |  6.71 ms | 27.41 dB |  145.90 kB
      20 | 26.91 ms | 32.70 dB |  697.20 kB |  6.74 ms | 30.32 dB |  198.30 kB
      30 | 27.17 ms | 34.63 dB |  850.60 kB |  6.84 ms | 31.92 dB |  243.60 kB
      40 | 27.19 ms | 35.97 dB |  958.90 kB |  6.89 ms | 32.99 dB |  282.20 kB
      50 | 27.29 ms | 36.94 dB | 1073.30 kB |  6.92 ms | 33.82 dB |  319.10 kB
      60 | 27.39 ms | 37.96 dB | 1217.10 kB |  6.95 ms | 34.65 dB |  360.00 kB
      70 | 27.51 ms | 39.22 dB | 1399.20 kB |  7.04 ms | 35.71 dB |  422.10 kB
      80 | 27.76 ms | 40.67 dB | 1710.00 kB |  7.13 ms | 37.15 dB |  526.70 kB
      90 | 28.36 ms | 42.83 dB | 2441.40 kB |  7.32 ms | 39.84 dB |  768.40 kB
     100 | 35.47 ms | 47.09 dB | 7798.70 kB |  9.31 ms | 47.21 dB | 2499.60 kB

Decoding:
         |           4k (4096x2160)         |         HD (1920x1080)
 --------+----------------------------------+---------------------------------
 quality | duration |     psnr |       size | duration |     psnr |       size
 --------+----------+----------+------------+---------------------------------	
      10 | 10.28 ms | 29.33 dB |  539.30 kB |  3.13 ms | 27.41 dB |  145.90 kB
      20 | 11.31 ms | 32.70 dB |  697.20 kB |  3.59 ms | 30.32 dB |  198.30 kB
      30 | 12.36 ms | 34.63 dB |  850.60 kB |  3.97 ms | 31.92 dB |  243.60 kB
      40 | 12.90 ms | 35.97 dB |  958.90 kB |  4.28 ms | 32.99 dB |  282.20 kB
      50 | 13.45 ms | 36.94 dB | 1073.30 kB |  4.56 ms | 33.82 dB |  319.10 kB
      60 | 14.71 ms | 37.96 dB | 1217.10 kB |  4.81 ms | 34.65 dB |  360.00 kB
      70 | 15.03 ms | 39.22 dB | 1399.20 kB |  5.24 ms | 35.71 dB |  422.10 kB
      80 | 16.64 ms | 40.67 dB | 1710.00 kB |  5.89 ms | 37.15 dB |  526.70 kB
      90 | 19.99 ms | 42.83 dB | 2441.40 kB |  7.48 ms | 39.84 dB |  768.40 kB
     100 | 46.45 ms | 47.09 dB | 7798.70 kB | 16.42 ms | 47.21 dB | 2499.60 kB
     
USAGE:
  1) LIBGPUJPEG LIBRARY:
      To build libgpujpeg library check REQUIREMENTS and go to 
    gpujpeg/libgpujpeg/ directory and run 'make' command. The shared library 
    object ./libgpujpeg.so will be build.
      To use library in your project you have to include library to your 
    sources and linked shared library object to your executable:
      
      #include "libgpujpeg/gpujpeg.h"
      
    ENCODING:
      For encoding by libgpujpeg library you have to declare two structures
    and set proper values to them. The first is definition of encoding/decoding 
    parameters, and the second is structure with parameters of input image:
      
      struct gpujpeg_parameters param;
      gpujpeg_set_default_parameters(&param);   
      param.quality = 80; 
      // (default value is 75)
      param.restart_interval = 16; 
      // (default value is 8)
      param.interleaved = 1;
      // (default value is 0)
      
      struct gpujpeg_image_parameters param_image;
      gpujpeg_image_set_default_parameters(&param_image);
      param_image.width = 1920;
      param_image.height = 1080;
      param_image.comp_count = 3;
      // (for now, it must be 3)
      param_image.color_space = GPUJPEG_RGB;
      // or GPUJPEG_BT709 or GPUJPEG_YCBCR_JPEG
      // (default value is GPUJPEG_RGB)
      param_image.pixel_format = GPUJPEG_444_U8_P012;
      // or eg. GPUJPEG_422_U8_P1020
      // (default value is GPUJPEG_444_U8_P012)
      
    If you want to use subsampling in JPEG format call following function,
    that will set default sampling factors (2x2 for Y, 1x1 for Cb and Cr):
    
      // Use default sampling factors
      gpujpeg_parameters_chroma_subsampling(&param);
      
    Or define sampling factors by hand:
    
      // User custom sampling factors
      param.sampling_factor[0].horizontal = 4;
      param.sampling_factor[0].vertical = 4;
      param.sampling_factor[1].horizontal = 1;
      param.sampling_factor[1].vertical = 2;
      param.sampling_factor[2].horizontal = 2;
      param.sampling_factor[2].vertical = 1;
        
    Next you have to initialize CUDA device by calling:
      
      if ( gpujpeg_init_device(device_id, 0) )
          return -1;
        
    where first parameters is CUDA device (e.g. device_id = 0) id and second 
    parameter is flag if verbose output should be used (0 or GPUJPEG_VERBOSE). 
    Next step is to create encoder:
    
      struct gpujpeg_encoder* encoder = gpujpeg_encoder_create(&param, 
          &param_image);
      if ( encoder == NULL )
          return -1;
        
    When creating encoder, library allocates all device buffers which will be 
    needed for image encoding and when you encode concrete image, they are 
    already allocated and encoder will used them for every image. Now we need 
    raw image data that we can encode by encoder, for example we can load it 
    from file:
    
      int image_size = 0;
      uint8_t* image = NULL;
      if ( gpujpeg_image_load_from_file("input_image.rgb", &image, 
               &image_size) != 0 )
          return -1;
            
    Next step is to encode uncompressed image data to JPEG compressed data 
    by encoder:
    
      struct gpujpeg_encoder_input encoder_input;
      gpujpeg_encoder_input_set_image(&encoder_input, image);
    
      uint8_t* image_compressed = NULL;
      int image_compressed_size = 0;
      if ( gpujpeg_encoder_encode(encoder, &encoder_input, &image_compressed, 
               &image_compressed_size) != 0 )
          return -1;
            
    Compressed data are placed in internal encoder buffer so we have to save 
    them somewhere else before we start encoding next image, for example we 
    can save them to file:
            
      if ( gpujpeg_image_save_to_file("output_image.jpg", image_compressed, 
               image_compressed_size) != 0 )
          return -1;
            
    Now we can load, encode and save next image or finish and move to clean up
    encoder. Finally we have to clean up so destroy loaded image and destroy 
    the encoder.
    
      gpujpeg_image_destroy(image);
      gpujpeg_encoder_destroy(encoder);
        
    DECODING:
      For decoding we don't need to initialize two structures of parameters. 
    We only have to initialize CUDA device if we haven't initialized it yet and 
    create decoder:
        
      if ( gpujpeg_init_device(device_id, 0) )
          return -1;
        
      struct gpujpeg_decoder* decoder = gpujpeg_decoder_create(NULL);
      if ( decoder == NULL )
          return -1;
            
    Now we have two options. The first is to do nothing and decoder will 
    postpone buffer allocations to decoding first image where it determines 
    proper image size and all other parameters. All the following images must 
    have the same parameters. The second option is to provide input image size 
    and optionally other parameters and the decoder will allocate all buffers 
    and it is fully ready when encoding even the first image. 
    
      struct gpujpeg_parameters param;
      gpujpeg_set_default_parameters(&param);
      param.restart_interval = 16; 
      param.interleaved = 1;
      
      struct gpujpeg_image_parameters param_image;
      gpujpeg_image_set_default_parameters(&param_image);
      param_image.width = 1920;
      param_image.height = 1080;
      param_image.comp_count = 3;
    
      // Pre initialize decoder before decoding
      gpujpeg_decoder_init(decoder, &param, &param_image);
      
    If you want to specify output image color space and/or subsampling factor,
    you can use following two parameters. You can specify them though the
    param structure befor passing it to gpujpeg_decoder_init. But if you
    postpone this initialization process to the first image, you have no
    other option than specify them in this way:
    
      gpujpeg_decoder_set_output_format(decoder, GPUJPEG_RGB,
                      GPUJPEG_444_U8_P012);
      // or eg. GPUJPEG_YCBCR_JPEG and GPUJPEG_422_U8_P1020
            
    Next we have to load JPEG image data from file and decoded it to raw 
    image data:
            
      int image_size = 0;
      uint8_t* image = NULL;
      if ( gpujpeg_image_load_from_file("input_image.jpg", &image, 
               &image_size) != 0 )
          return -1;
    
      struct gpujpeg_decoder_output decoder_output;
      gpujpeg_decoder_output_set_default(&decoder_output);
      if ( gpujpeg_decoder_decode(decoder, image, image_size, 
               &decoder_output) != 0 )
          return -1;
            
    Now we can save decoded raw image data to file and perform cleanup:
            
      if ( gpujpeg_image_save_to_file("output_image.rgb", decoder_output.data, 
               decoder_output.data_size) != 0 )
          return -1;
        
      gpujpeg_image_destroy(image);
      gpujpeg_decoder_destroy(decoder);
      
    
  2) GPUJPEG CONSOLE APPLICATION:
      The console application gpujpeg uses libgpujpeg library to demonstrate
    it's functions. To build console application check REQUIREMENTS and go to
    gpujpeg directory (where README and LICENSE files are placed) and run 
    'make' command. It builds libgpugjpeg library in subdirectory 
    ./libgpujpeg/ and it creates executable file ./gpujpeg and run script 
    ./gpujpeg.sh, which runs executable file linked to runtime library 
    libgpujpeg.so (which is placed in ./libgpujpeg/ subdirectory).
      To encode image from raw RGB image file to JPEG image file use following
    command:
      
      ./gpujpeg.sh --encode --size=WIDTHxHEIGHT --quality=QUALITY \ 
              INPUT_IMAGE.rgb OUTPUT_IMAGE.jpg
      
    You must specify input image size by --size=WIDTHxHEIGHT parameter. 
    Optionally you can specify desired output quality by parameter 
    --quality=QUALITY which accepts values 0-100. Console application accepts 
    a few more parameters and you can list them by folling command:
    
      ./gpujpeg.sh --help
      
    To decode image from JPEG image file to raw RGB image file use following 
    command:
      
      ./gpujpeg.sh --decode OUTPUT_IMAGE.jpg INPUT_IMAGE.rgb
      
    You can also encode and decode image to test the console application:
      
      ./gpujpeg.sh --encode --decode --size=WIDTHxHEIGHT --quality=QUALITY \
              INPUT_IMAGE.rgb OUTPUT_IMAGE.jpg
      
    Decoder will create new decoded file OUTPUT_IMAGE.jpg.decoded.rgb and do 
    not overwrite your INPUT_IMAGE.rgb file.
      Console application is able to load raw RGB image file data from *.rgb 
    files and raw YUV and YUV422 data from *.yuv files. For YUV422 you must 
    specify *.yuv file and use '--sampling-factor=4:2:2' parameter. 
    All supported parameters for console application are following:
      --help
          Prints console application help
      --size=1920x1080
          Input image size in pixels, e.g. 1920x1080
      --sampling-factor=4:4:4
          Input image sampling factor (supported are '4:4:4' and '4:2:2')
      --colorspace=rgb
          Input image colorspace (supported are 'rgb', 'yuv' and 'ycbcr-jpeg', 
          where 'yuv' means YCbCr ITU-R BT.601), when *.yuv file is specified,
          instead of default 'rgb', automatically the colorspace 'yuv' is used
      --quality
          Set output quality level 0-100 (default 75)
      --restart=8
          Set restart interval for encoder, number of MCUs between 
          restart markers
      --subsampled
          Produce chroma subsampled JPEG stream
      --interleaved
          Produce interleaved stream
      --encode
          Encode images
      --decode
          Decode images
      --device=0
          By using this parameter you can specify CUDA device id which will 
          be used for encoding/decoding.
    
    Restart interval is important for parallel huffman encoding and decoding.
    When '--restart=N' is used (default is 8), the coder can process each 
    N MCUs independently, and so he can code each N MCUs in parallel. When 
    '--restart=0' is specified, restart interval is disabled and the coder 
    must use CPU version of huffman coder (because on GPU would run only one 
    thread, which is very slow).
      The console application can encode/decode multiple images by following 
    command:
      
      ./gpujpeg.sh ARGUMENTS INPUT_IMAGE_1.rgb OUTPUT_IMAGE_1.jpg \
              INPUT_IMAGE_2.rgb OUTPUT_IMAGE_2.jpg ...
    
REQUIREMENTS:
    To be able to build and run libgpujpeg library and gpujpeg console 
  application you need:
  1) CUDA Toolkit (http://developer.nvidia.com/cuda-toolkit) installed,
     default installation path is /usr/local/cuda. If you have the CUDA 
     installed somewhere else, you need to specify it by environment variable 
     CUDA_INSTALL_PATH or in Makefiles by CUDA_INSTALL_PATH variable.
  2) NVIDIA developer drivers
  3) CUDA enabled NVIDIA GPU

LICENSE:
  See file LICENSE.
  This software contains source code provided by NVIDIA Corporation.
  This software source code is based on SiGenGPU [3].

REFERENCES:
  [1] http://www.w3.org/Graphics/JPEG/itu-t81.pdf
  [2] http://www.ijg.org/
  [3] https://github.com/silicongenome/SiGenGPU
  [4] http://www.ecma-international.org/publications/files/ECMA-TR/TR-098.pdf
  


NEWS
2020-10-02
----------
- Support for SPIFF files (decode, encode limited-range BT.709, BT.601)
- Support for FFmpeg limited-range BT.601 JPEGs

2020-05-12
----------
- Support for PNM and PAM files

2020-03-10
----------
- Command-line application is now supported in MSW

2020-01-21
----------
- Support for RGBA buffers
- Raw RGBA and I420 files supported

2019-12-02
----------
- Decoder compatibility with more (foreign) JPEGs

2019-06-18
----------
- Support for RGB JPEG images

2016-03-07
----------
- Support for decoding to OpenGL texture or CUDA buffer
- Encoding different image sizes
- Support for encoding of planar images (limited)

2013-03-19
----------
- New IDCT computing kernel 43 % faster

2012-08-02
----------
Changes:
-Added new parallel GPU Huffman encoder for compute capabilities >= 2.0
-Rewritten GPU Huffman decoder
-Rewritten GPU forward DCT, preserving precision of fixed point implementation 
 and gaining better performance than NPP implementation.
 (Both old implementations were thus removed.)
-Rewritten GPU encoder preprocessor to remove unnecessary operations
-Minor performance improvement in GPU encoder postprocessor

2012-03-29
----------
Changes:
-Optionally the encoder input or decoder output can be loaded/store
 from/to OpenGL texture.

2012-03-09
----------
Changes:
-Refactored color spaces transformations, added new color spaces
-Added new implementation of DCT/IDCT on CUDA, it is slightly slower
 than the NPP implementation, but the new IDCT doesn't cause color space 
 change. It can be turned on by Makefile option. By default is NPP 
 version used.

2012-02-24
----------
Changes:
-Slightly improved performance of encoder.
-Correction of segment info for large images

2012-02-21
----------
Changes:
-Added option to encoder for generating segment info into APP13 header that is used
 by decoder to perform fast stream parsing.
-Renamed --interleving option to --interleaved and --chroma-subsampling to --subsampled.

2012-02-16
----------
Changes:
-Added option for verbose output from console application.
-Library now stores all coder time durations inside coder structure and it
 can be accessed from userspace.
-Console application now prints GPU allocation info (when --verbose is used).
-Correction of encoding/decoding for large images (e.g. 4320p or 8K)

2012-02-07
----------
-Improved preprocessor and postprocessor performance.
-Added recovering from error inside JPEG stream to decoder.
-Library provides function for listing CUDA devices.
-Library provides option for decoder to output result right into OpenGL PBO resource
 which improves performance when the result should be displayed on the same GPU.
 -Correction of build warnings.
 -Correction of other errors.

2012-01-04
----------
Changes:
-Added option to libgpujpeg Makefile for moving gpu huffman coder tables into constant memory.
-Option is default set to true and makes better performance on older GPUs.

2012-01-03
----------
Changes:
-Added interleaved mode (optionally).
-Added subsampling (optionally) - better performance when used 
 (because of smaller amount of encoded/decoded data).

2011-12-07
----------
First public release of gpujpeg library and console 
application.
Features:
-Baseline JPEG, non-interleaved mode on CUDA-enabled GPUs.
-Performance - realtime HD and 4K encoding/decoding on NVIDIA GTX 580.
-Using CUDA and NPP library.


jpeg_table.h
/**
 * Copyright (c) 2011, Martin Srom
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef JPEG_TABLE
#define JPEG_TABLE

#include "jpeg_type.h"

/** JPEG natural order from zigzag order */
static const int jpeg_order_natural[64] = {
     0,  1,  8, 16,  9,  2,  3, 10,
    17, 24, 32, 25, 18, 11,  4,  5,
    12, 19, 26, 33, 40, 48, 41, 34,
    27, 20, 13,  6,  7, 14, 21, 28,
    35, 42, 49, 56, 57, 50, 43, 36,
    29, 22, 15, 23, 30, 37, 44, 51,
    58, 59, 52, 45, 38, 31, 39, 46,
    53, 60, 61, 54, 47, 55, 62, 63
};

/** JPEG quantization table structure */
struct jpeg_table_quantization
{
    // Quantization raw table
    uint8_t table_raw[64];
    // Quantization forward/inverse table
    uint16_t table[64];
    // Quantization forward/inverse table in device memory
    uint16_t* d_table;
};

/** JPEG table for huffman encoding */
struct jpeg_table_huffman_encoder {
    // Code for each symbol 
    unsigned int code[256];	
    // Length of code for each symbol 
	char size[256];
    // If no code has been allocated for a symbol S, size[S] is 0 

	// These two fields directly represent the contents of a JPEG DHT marker
    // bits[k] = # of symbols with codes of length k bits; bits[0] is unused
    unsigned char bits[17];
    // The symbols, in order of incr code length
    unsigned char huffval[256];
};

/** JPEG table for huffman decoding */
struct jpeg_table_huffman_decoder {
    // Smallest code of length k
    int mincode[17]; 
    // Largest code of length k (-1 if none) 
	int maxcode[18];
    // Huffval[] index of 1st symbol of length k
	int valptr[17];
    // # bits, or 0 if too long
    int look_nbits[256];
    // Symbol, or unused
	unsigned char look_sym[256];
    
    // These two fields directly represent the contents of a JPEG DHT marker
    // bits[k] = # of symbols with codes of 
	unsigned char bits[17];
    // The symbols, in order of incr code length 
	unsigned char huffval[256];
};

/**
 * Init JPEG quantization table for encoder
 * 
 * @param table  Table structure
 * @param type  Type of component for table
 * @param quality  Quality (0-100)
 * @return 0 if succeeds, otherwise nonzero
 */
int
jpeg_table_quantization_encoder_init(struct jpeg_table_quantization* table, enum jpeg_component_type type, int quality);

/**
 * Init JPEG quantization table for decoder
 * 
 * @param table  Table structure
 * @param type  Type of component for table
 * @param quality  Quality (0-100)
 * @return 0 if succeeds, otherwise nonzero
 */
int
jpeg_table_quantization_decoder_init(struct jpeg_table_quantization* table, enum jpeg_component_type type, int quality);

/**
 * Compute JPEG quantization table for decoder
 * 
 * @param table  Table structure
 * @return 0 if succeeds, otherwise nonzero
 */
int
jpeg_table_quantization_decoder_compute(struct jpeg_table_quantization* table);

/**
 * Print JPEG quantization table
 * 
 * @param table  Table structure
 * @return void
 */
void
jpeg_table_quantization_print(struct jpeg_table_quantization* table);

/**
 * Initialize encoder huffman DC and AC table for component type
 * 
 * @param table  Table structure
 * @param comp_type  Component type (luminance/chrominance)
 * @param huff_type  Huffman type (DC/AC)
 * @return void
 */
int
jpeg_table_huffman_encoder_init(struct jpeg_table_huffman_encoder* table, struct jpeg_table_huffman_encoder* d_table, enum jpeg_component_type comp_type, enum jpeg_huffman_type huff_type);

/**
 * Initialize decoder huffman DC and AC table for component type. It copies bit and values arrays to table and call compute routine.
 * 
 * @param table  Table structure
 * @param comp_type  Component type (luminance/chrominance)
 * @param huff_type  Huffman type (DC/AC)
 * @return void
 */
int
jpeg_table_huffman_decoder_init(struct jpeg_table_huffman_decoder* table, enum jpeg_component_type comp_type, enum jpeg_huffman_type huff_type);

/** 
 * Compute decoder huffman table from bits and values arrays (that are already set in table)
 * 
 * @param table
 * @return void
 */
void
jpeg_table_huffman_decoder_compute(struct jpeg_table_huffman_decoder* table);

#endif // JPEG_TABLE